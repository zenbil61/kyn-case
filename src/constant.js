const ROUTE = {
  PAGE_JOBS: '/jobs',
  PAGE_JOBS_DETAIL: '/jobs/:id',
};

const ACTIONS = {
  FETCH_JOB: 'fetchJob',
  FETCH_JOBS: 'fetchJobs',
  FETCH_CITIES: 'fetchCities',
};

const MUTATIONS = {
  SET_JOB: 'setJob',
  SET_JOB_LOADING: 'setJobLoading',
  SET_JOBS: 'setJobs',
  SET_JOBS_LOADING: 'setJobsLoading',
  SET_CITIES: 'setCities',
};

const CITIES = [
  { text: 'İstanbul', value: 34 },
  { text: 'Ankara', value: 6 },
  { text: 'İzmir', value: 35 },
];
export { ROUTE, ACTIONS, MUTATIONS, CITIES };
