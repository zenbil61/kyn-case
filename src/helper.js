function paginator(items, current_page, per_page_items) {
  let page = current_page || 1;
  let per_page = per_page_items || 10;
  let offset = (page - 1) * per_page;
  let paginatedItems = items.slice(offset).slice(0, per_page_items);
  return paginatedItems;
}
function wait(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
const debounce = (func, wait) => {
  let timeout;

  return function executedFunction(...args) {
    const later = () => {
      timeout = null;

      func(...args);
    };

    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  };
};

export { paginator, debounce, wait };
