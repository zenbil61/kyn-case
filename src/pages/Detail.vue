<template>
  <div v-if="job_loading">YÃ¼kleniyor...</div>
  <div v-if="!job_loading">
    <Button @click="handleBack">Geri</Button>
    <br />
    <br />
    <h2>{{ job.title }}</h2>
    <br />
    <p>{{ job.area }} / {{ cityName }}</p>
    <p>{{ job.description }}</p>
  </div>
</template>

<script>
export default {
  name: 'DetailComp',
};
</script>
<script setup>
import { computed, onMounted } from 'vue';
import { useRoute, useRouter } from 'vue-router';
import { Button } from '../components/ui';
import { useStore } from 'vuex';
import { ACTIONS } from '../constant';
const store = useStore();
const { params } = useRoute();
const router = useRouter();

// computed
const job = computed(() => store.state.job);
const job_loading = computed(() => store.state.job_loading);
const cityName = computed(() => store.getters.cityName(job?.value?.city || ''));

// events
const handleBack = () => {
  store.dispatch('clearJob');
  router.go(-1);
};

onMounted(() => {
  const hasCities = !store.state.cities || store.state.cities.length < 1;
  if (hasCities) {
    router.push('/');
    return;
  }

  store.dispatch(ACTIONS.FETCH_JOB, { id: params.id });
});
</script>
