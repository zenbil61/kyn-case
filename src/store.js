import { createStore } from 'vuex';
import { ACTIONS, MUTATIONS } from './constant';
import axios from 'axios';
import { wait } from './helper';
const BASE_API_URL = process.env.VUE_APP_BASE_API_URL;

const store = createStore({
  state() {
    return {
      jobs_data: {},
      jobs_data_loading: true,
      cities: [],
      job: {},
      job_loading: true,
    };
  },
  mutations: {
    [MUTATIONS.SET_JOBS]: (state, data) => (state.jobs_data = data),
    [MUTATIONS.SET_JOBS_LOADING]: (state, data) => (state.jobs_data_loading = data),
    [MUTATIONS.SET_JOB]: (state, data) => (state.job = data),
    [MUTATIONS.SET_JOB_LOADING]: (state, data) => (state.job_loading = data),
    [MUTATIONS.SET_CITIES]: (state, data) => (state.cities = data),
  },
  actions: {
    async [ACTIONS.FETCH_JOBS]({ commit }, request) {
      commit(MUTATIONS.SET_JOBS_LOADING, true);
      await wait(500);
      const { data } = await axios.get(`${BASE_API_URL}/jobs`, {
        params: request,
      });
      commit(MUTATIONS.SET_JOBS, data.data);
      commit(MUTATIONS.SET_JOBS_LOADING, false);
    },
    async [ACTIONS.FETCH_CITIES]({ commit }) {
      const { data } = await axios.get(`${BASE_API_URL}/cities`);
      commit(MUTATIONS.SET_CITIES, data.data);
    },
    async [ACTIONS.FETCH_JOB]({ commit }, { id }) {
      commit(MUTATIONS.SET_JOB_LOADING, true);
      await wait(500);

      const { data } = await axios.get(`${BASE_API_URL}/jobs/detail`, {
        params: { id },
      });
      commit(MUTATIONS.SET_JOB, data.data);
      commit(MUTATIONS.SET_JOB_LOADING, false);
    },
  },
  getters: {
    cityName: (state) => (id) => {
      return state.cities.find((city) => city.value == id).text;
    },
  },
});

export default store;
