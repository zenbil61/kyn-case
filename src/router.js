import { createRouter, createWebHistory } from 'vue-router';

import Home from './pages/Home.vue';
import Detail from './pages/Detail.vue';
import List from './pages/List.vue';

import { ROUTE } from './constant';
const routes = [
  { path: '/', component: Home },
  { path: ROUTE.PAGE_JOBS, component: List },
  {
    path: ROUTE.PAGE_JOBS_DETAIL,
    component: Detail,
  },
];
const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;
