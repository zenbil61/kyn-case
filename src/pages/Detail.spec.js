import { mount } from '@vue/test-utils';
import Detail from './Detail.vue';
import router from '@/router';
import store from '@/store';
import { wait } from '@/helper';
import { CITIES } from '@/constant';

it('renders correctly', async () => {
  store.state.cities = CITIES;
  router.currentRoute.value.params.id = 0;

  const wrapper = mount(Detail, {
    global: {
      plugins: [router, store],
    },
    props: {},
  });
  wait(1500);

  expect(wrapper.element).toMatchSnapshot();
});
