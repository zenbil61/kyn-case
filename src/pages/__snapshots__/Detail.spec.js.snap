// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <div>
    YÃ¼kleniyor...
  </div>
  <!--v-if-->
  
</div>
`;
